src/vmq_reg.erl

```diff
@@ -636,88 +604,54 @@ fold_sessions(FoldFun, Acc) ->
 
 -spec add_subscriber([{topic(), qos() | not_allowed}], subscriber_id()) -> ok.
 add_subscriber(Topics, SubscriberId) ->
-    Node = node(),
-    OldSubs = plumtree_metadata:get(?SUBSCRIBER_DB, SubscriberId, [{default, []}]),
-    {NewSubs, HasChanged} =
-    lists:foldl(fun ({_Topic, not_allowed}, {NewSubsAcc, Changed}) ->
-                        {NewSubsAcc, Changed};
-                    ({Topic, QoS}, {NewSubsAcc, Changed}) ->
-                        NewSub = {Topic, QoS, Node},
-                        case lists:keyfind(Topic, 1, NewSubsAcc) of
-                            NewSub ->
-                                %% exactly the same subscription
-                                %% ignore it
-                                {NewSubsAcc, Changed};
-                            {Topic, _, _Node} ->
-                                %% same topic filter, but different qos
-                                %% replace subscription: [MQTT-3.8.4-3]
-                                {lists:keyreplace(Topic, 1, NewSubsAcc,
-                                                  NewSub), true};
-                            false ->
-                                %% new subscription
-                                {[NewSub|NewSubsAcc], true}
-                        end
-                end, {OldSubs, false}, Topics),
-    case HasChanged of
-        true ->
-            %% only store if something changed
-            plumtree_metadata:put(?SUBSCRIBER_DB, SubscriberId, NewSubs);
-        false ->
+    OldSubs = subscriptions_for_subscriber_id(SubscriberId),
+    case vmq_subscriber:add(OldSubs, [{T, QoS}||{T, QoS} <- Topics, is_integer(QoS)]) of
+        {NewSubs, true} ->
+            vmq_subscriber_db:store(SubscriberId, NewSubs);
+        _ ->
```

Comment:
This change I wasn't able to follow - it's not clear to me that the removed fold above and is equivalent to the `vmq_subscriber:add/2`. I guess I need to spend more time understanding the new code there.